Add shadow support.

donmccurdy opened the pull request:
**Description:**

Fixes #651.

Migrates [shadows from aframe-extras](https://github.com/donmccurdy/aframe-extras/tree/master/src/shadows), with a few changes to the API. Example includes shadows on the "Hello World" example, but we should probably come up with something more interesting.

Some of the shadow properties are non-default; this has worked better so far (from various bugs filed against extras) but perhaps we want to be consistent with THREE defaults instead? In any case, tweaks to the shadow camera properties are going to be par for the course on any serious scene.

**Changes proposed:**
- Add `shadow` component, which recursively enables cast/receive on child geometry. Note that it is recursive and on by default.
- Add shadow properties to the `light` component, for supported types.
- Add shadow setup to the `light` system. The shadowMap is enabled automatically, and not exposed in schema, but the shadow type can be set with `<a-scene light="shadowMapType: pcfsoft">`.

/cc @fernandojsg @takahirox

donmccurdy commented on src/components/light.js:
```diff
@@ -81,6 +95,21 @@ module.exports.Component = registerComponent('light', {
+          case 'shadowMapHeight':
+          case 'shadowMapWidth':
+            this.updateShadow();
+            break;
```
Hm this gets executed repeatedly. I'll try to think of a better way to do this part.

donmccurdy replied:
Done.

takahirox commented:
`renderSingleSided` and `renderReverseSided` I requested aren't supported yet?

https://github.com/donmccurdy/aframe-extras/issues/117

donmccurdy commented:
Oops I forget. Added!

takahirox commented:
Thanks!

takahirox commented on src/components/shadow.js:
```diff
@@ -0,0 +1,43 @@
+    this.el.object3D.traverse(function (node) {
+      if (node instanceof THREE.Mesh) {
+        node.castShadow = data.cast;
+        node.receiveShadow = data.receive;
```
I think we need to update material here and in remove() like this
to handle set/removeAttribute('shadow') executed after scene is rendered.

    var material = node.material;
    if (material.isMultlMaterial === true) {
        for (var i = 0, il = material.materials.length; i < il; i++) {
            material.materials[i].needsUpdate = true;
        }
    } else {
        material.needsUpdate = true;
    }

Reference:
Run the following commands one by one in the console at
https://threejs.org/examples/webgl_animation_skinning_morph.html

    mesh.receiveShadow = false
    mesh.material.materials[0].needsUpdate = true;
    mesh.receiveShadow = true
    mesh.material.materials[0].needsUpdate = true;

donmccurdy replied:
(see below — is this just for THREE.MultiMaterial?)

takahirox replied:
No. Not MultiMaterial specific, but general material specification.

takahirox commented:
Just curious to know if we can dynamically enable/disable shadow after scene is rendered once?

donmccurdy commented:
> Just curious to know if we can dynamically enable/disable shadow after scene is rendered once?

Programmatically, yes: `sceneEl.systems.light.setShadowMapEnabled(false)`. Systems don't support updates right now, so there's no pure-HTML way.

takahirox commented:
I see, supporting update is our next step :D

takahirox commented on src/systems/light.js:
```diff
@@ -73,5 +98,17 @@ module.exports.System = registerSystem('light', {
+    this.shadowMapEnabled = enabled;
+    if (renderer) {
+      renderer.shadowMap.enabled = enabled;
+    }
```
You need to update materials in a scene to dynamically handle shadow on/off 
after scene is rendered once if you switch `shadowMap.enabled`.

Reference:
Run the following commands one by one on the console at
https://threejs.org/examples/webgl_animation_skinning_morph.html

    renderer.shadowMap.enabled = false;
    mesh.material.materials[0].needsUpdate = true;
    renderer.shadowMap.enabled = true;
    mesh.material.materials[0].needsUpdate = true;

donmccurdy replied:
I don't see this in most shadow demos/docs, is there a particular type of material this effects?

takahirox replied:
This is a general material issue.

This is necessary if we dynamically turn on/off shadow after scene is rendered once.
So it's kinda rare situation and maybe it isn't written in any documents.

Shadow related GLSL code is kinda hard coded depending on shadow parameters.
So we need to set `material.needsUpdate` `true` to update GLSL code
if we dynamically switch `shadowMap.enable` on/off.

Probably they'll help you understand.
https://github.com/mrdoob/three.js/blob/93bfc79da1758d4891f27f809598f836728b009b/src/renderers/webgl/WebGLPrograms.js#L190
https://github.com/mrdoob/three.js/blob/ddd8c2b1eb8a92a344b261c2125d7a2997248309/src/renderers/webgl/WebGLProgram.js#L358

donmccurdy replied:
Thanks, makes sense. Updated!

takahirox replied:
Cool! And I found this's written in the Three.js document.

https://threejs.org/docs/#Manual/Next_Steps/How_to_update_things

takahirox commented:
OK, I see.

I inserted some comments inline to support dynamic updates just for your information.

dmarcos commented on src/components/light.js:
```diff
@@ -20,7 +20,21 @@ module.exports.Component = registerComponent('light', {
-    target: {type: 'selector', if: {type: ['spot', 'directional']}}
+    target: {type: 'selector', if: {type: ['spot', 'directional']}},
+
+    // Shadows.
```
Should we make the light a component with dynamic schema if a the property `castShadows: true`?

donmccurdy replied:
Either way IMO. Preference?

dmarcos commented on src/components/shadow.js:
```diff
@@ -0,0 +1,43 @@
+    this.el.addEventListener('model-loaded', this.onMeshChanged);
+  },
+
+  update: function () {
```
What happens if we add entities dynamically to the scene? Should we listen for the `loaded` event?

dmarcos replied:
disregard. I thought this was the system

dmarcos replied:
but still. A mesh could be added to the entity without being a model. Should we listen for `'object3dset' to update the shadow flags?

donmccurdy replied:
`object3dset` sounds like a more reliable option, sure.

donmccurdy replied:
Done.

ngokevin commented:
Any tips on tuning the shadows?

**Single Point Light + Torus Knot**

Point light in the center of the knot. Boxy shadow shapes.

<img width="647" alt="screen shot 2017-02-06 at 6 56 00 pm" src="https://cloud.githubusercontent.com/assets/674727/22676034/f780c2fc-ec9d-11e6-9fe6-cdd29083338a.png">

**Directional Light + Point Light**

Trying to animate the point light, spinning around the scene. Point light casts odd shape for the cylinder at an odd angle. Do I have to tweak `light.angle`?

<img width="856" alt="screen shot 2017-02-06 at 6 53 06 pm" src="https://cloud.githubusercontent.com/assets/674727/22675970/83b43354-ec9d-11e6-861a-e3e10dfe36fa.png">

ngokevin commented on src/systems/light.js:
```diff
@@ -14,13 +21,33 @@ var DEFAULT_LIGHT_ATTR = 'data-aframe-default-light';
  */
 module.exports.System = registerSystem('light', {
+  schema: {
+    shadowMapType: {default: 'pcf', oneOf: ['basic', 'pcf', 'pcfsoft']},
```
This seems like it should be a shadow system to match the shadow component?

```shadow="type: basic"```

donmccurdy replied:
Semantically nice. So then shadows would be enabled on the renderer if any `shadow` components are present, not based on whether `light="castShadow: true"` is set?

donmccurdy replied:
Done.

ngokevin commented:
I continued playing around. Got it looking a bit better, but doesn't look as smooth as the [three.js shadow + point light example](https://threejs.org/examples/#webgl_shadowmap_pointlight), even though I've cloned all the parameters.

https://github.com/ngokevin/aframe/commit/958fdfc2b2cfe37508951dbc439f88ad81dc4cab

<img width="965" alt="screen shot 2017-02-07 at 3 30 08 pm" src="https://cloud.githubusercontent.com/assets/674727/22716319/53c0d32a-ed4a-11e6-8e2d-5afd2ddf58fd.png">

I also added some meshes (sphere, box, cylinder), the shadows seem clipped and in a weird spot despite playing the shadow camera top/near/left/right and map width/height

donmccurdy commented:
Re: tuning, I think scene dimensions are very much a contributor to how your camera parameters should be set. Not really sure how to optimize that. I've just pushed a `shadowCameraVisible` option that is helpful for debugging.

Directional lights will perform a fair bit better than point lights, as well.

ngokevin commented:
Still playing with the shadows. The camera helper really helps, I saw the camera was totally not where I expected it to be. Should we start adding baseline tests and docs?

donmccurdy commented:
Tests added.

ngokevin commented:
Was trying to get rid of some shadow clipping and debugging why my shadow camera was not lined up with my light, but playing with the shadow parameters from the Inspector doesn't have an effect. Are the updates working?

ngokevin commented:
Will save this for 0.5.1 until I get a good handle on it and document the tuning process.

donmccurdy commented:
Sounds good. I've fixed the updates — there were a couple more hoops to jump through for updating the shadow camera and cameraHelper. Also removed `shadowDarkness`, which isn't supported anymore.

donmccurdy commented:
Added a first draft of documentation. I'm not attached to the structure really: there's no *Systems* category in the docs, and most of the properties are on the `light` component, so that's where I put most of the documentation. Perhaps a post in the *Guides* section would be more appropriate?

We should probably also link to something explaining light baking, but I'm just learning that process myself.

donmccurdy commented:
Rebased, and reverted temporary changes to the hello-world example.

Any more to do here? Should an example be part of this PR?

ngokevin commented:
Thanks, this is my next priority. I have that example I was doing to copy the threejs example, if you have any cool ones we can add that too

ngokevin commented:
We can cherry-pick this example for now: https://github.com/ngokevin/aframe/commit/958fdfc2b2cfe37508951dbc439f88ad81dc4cab

I had *much* better and expected results setting `shadowCameraFov` to 90. With the help of the Inspector.

ngokevin commented:
I added an example, changed defaultCameraFov to 90 for better point lights.

Dynamic schemas we could do in a follow up.
